### PR TITLE
Wrote unit test for Hilbert transform

### PR DESCRIPTION
**Pull request by dpaulzc**
Moved the Hilbert transform subroutine to misc.f90 and wrote a unit test for the same.

### CODE REVIEW

File: src/misc.f90 [hunk 1]
```diff
@@ -1653,4 +1653,49 @@ subroutine subtitle(text)
...
+
+  subroutine Hilbert_transform(fx, Hfx)
+     !! Does Hilbert tranform for a given function
+     !! Ref - EQ (4)3, R. Balito et. al.
+     !! - An algorithm for fast Hilbert transform of real
+     !!   functions
+     !!
+     !! fx is the function
+     !! Hfx is the Hilbert transform of the function
+     !!
```
Comment (nakib) [inline src/misc.f90]: Delete this random line.

File: src/misc.f90 [hunk 2]
```diff
@@ -1653,4 +1653,49 @@ subroutine subtitle(text)
...
+
+     real(r64), allocatable, intent(out) :: Hfx(:)
+     real(r64), intent(in) :: fx(:)
```
Comment (nakib) [inline src/misc.f90:1662]: I prefer to have the `intent(in)` arguments declared before the `intent(out)` arguments.

File: src/misc.f90 [hunk 3]
```diff
@@ -1653,4 +1653,49 @@ subroutine subtitle(text)
...
+     real(r64) :: term2, term3, b
+
+     nfx = size(fx)
+     allocate(Hfx(nfx))
+
+     ! Hilbert function is zero at the edges
+     Hfx(1) = 0.0_r64
+     Hfx(nfx) = 0.0_r64
+
+     ! Both Hfx and fx follow 1-indexing system unlike the paper
```
Comment (nakib) [inline src/misc.f90]: This comment is unclear. What do you mean by this? Perhaps you want to add a note like "Note that in the reference, we have N + 1 points and the indexing is 0-based. On the other hand, here we have N points and the indexing is 1-based."

File: src/misc.f90 [hunk 4]
```diff
@@ -1653,4 +1653,49 @@ subroutine subtitle(text)
...
+     do k = 1, nfx - 2      ! Run over the internal points
```
Comment (nakib) [inline src/misc.f90]: Why is there a random amount of white space between the end of the line and the beginning of the comment?

File: src/misc.f90 [hunk 5]
```diff
@@ -1653,4 +1653,49 @@ subroutine subtitle(text)
...
+        term2 = 0.0_r64  ! 2nd term in Bilato Eq. 4
+        term3 = 0.0_r64  ! 3rd term in Bilato Eq. 4
+
+        do n = 1, nfx - 2 - k  ! Partial sum over internal points
+           b = log((n + 1.0_r64)/n)
+           term2 = term2 - (1.0_r64 - (n + 1.0_r64)*b)*fx(k + n + 1) + &
+                          (1.0_r64 - n*b)*fx(k + n + 2)
+        end do
+
+        do n = 1, k - 1      ! Partial sum over internal points
```
Comment (nakib) [inline src/misc.f90]: Why is there a random amount of white space between the end of the line and the beginning of the comment?

File: src/misc.f90 [hunk 6]
```diff
@@ -1653,4 +1653,49 @@ subroutine subtitle(text)
...
+           b = log((n + 1.0_r64)/n)
+           term3 = term3 + (1.0_r64 - (n + 1.0_r64)*b)*fx(k - n + 1) - &
+                         (1.0_r64 - n*b)*fx(k - n)
+        end do
+
+        Hfx(k + 1) = (-1.0_r64/pi)*(fx(k + 2) - fx(k) + term2 + term3)
```
Comment (nakib) [inline src/misc.f90]: Why not just write `Hfx(k + 1) = -(fx(k + 2) - fx(k) + term2 + term3)/pi`. I don't see the point of doing an extra arithmetic operation.

File: test/test_misc.f90 [hunk 7]
```diff
@@ -334,8 +337,87 @@ program test_misc
...
+     end function fx2_array
+
+     ! Hfx2 is actual hilbert transform of fx2, is an even function
+     function hfx2_array(x) result(hfx2)
+        real(r64), intent(in) :: x(:)
+        real(r64), allocatable :: hfx2(:)
+        real(r64), parameter :: e = 2.718281
+        allocate(hfx2(size(x)))
+        hfx2 = (1/e - cos(x))/(1 + x**2)
+     end function hfx2_array
```
Comment (nakib) [inline test/test_misc.f90]: This is ok, but inelegant. As discussed earlier, write an `elemental` function.

File: test/test_misc.f90 [hunk 8]
```diff
@@ -334,8 +337,87 @@ program test_misc
...
+        real(r64), allocatable :: fx2(:)
+        allocate(fx2(size(x)))
+        fx2 = sin(x)/(1 + x**2)
+     end function fx2_array
+
+     ! Hfx2 is actual hilbert transform of fx2, is an even function
+     function hfx2_array(x) result(hfx2)
+        real(r64), intent(in) :: x(:)
+        real(r64), allocatable :: hfx2(:)
+        real(r64), parameter :: e = 2.718281
```
Comment (nakib) [inline test/test_misc.f90]: Why do you need to hard code the natural constant, that too only up to 6 decimal places? You can simply use `exp(1.0_r64)` when needed.

File: test/test_misc.f90 [hunk 9]
```diff
@@ -334,8 +337,87 @@ program test_misc
...
+        allocate(hfx2(size(x)))
+        hfx2 = (1/e - cos(x))/(1 + x**2)
```
Comment (nakib) [inline test/test_misc.f90]: Check if you are having loss of precision by adding an integer to a `r64` here and elsewhere.

File: test/test_misc.f90 [hunk 10]
```diff
@@ -334,8 +337,87 @@ program test_misc
...
+
+  itest = itest + 1
+  test_array(itest) = testify("Hilbert transform: odd function, even points")
+  allocate(hfx2_even(n_even))
+  call Hilbert_transform(fx2_array(x_even), hfx2_even)
+  call test_array(itest)%assert(hfx2_even(ind_even), hfx2_array(x_even(ind_even)), &
+       tol = 1e-4_r64)
+
+  itest = itest + 1
+  test_array(itest) = testify("Hilbert transform: even function, odd points")
```
Comment (nakib) [inline test/test_misc.f90]: Write the function instead of just saying even/odd function here and elsewhere.

File: test/test_misc.f90 [hunk 11]
```diff
@@ -334,8 +337,87 @@ program test_misc
   call shrink(array_of_reals, 2_i64)
   call test_array(itest)%assert(array_of_reals, [1, 2]*1.0_r64)
   
+  ! Hilbert transform tests (H)
+  ! fx1 -> function 1, fx2 -> function 2
+  ! - hfx1_even stores hilbert transform calculated for fx1, and for even number
```
Comment (nakib) [inline test/test_misc.f90]: What does this `-` mean?

File: test/test_misc.f90 [hunk 12]
```diff
@@ -334,8 +337,87 @@ program test_misc
...
+  ! - of points
```
Comment (nakib) [inline test/test_misc.f90]: What does this `-` mean? Seems totally random.

**Review by nakib — CHANGES_REQUESTED** [review]
Hi Dwaipayan,

Thanks a lot for the PR. I left some comments below. Please resolve all of them.

Best,
Nakib

File: test/test_misc.f90 (hunk 10 above)
Comment (dpaulzc) [reply test/test_misc.f90]: Do I write like - ```"Hilbert transform; function: sinx/(1+x^2), odd points"``` ?

Comment (nakib) [reply test/test_misc.f90]: Something like: "Hilbert transform: f(x) = sin(x)/(1 + x^2), odd points" would be nice. Please use proper spacing around operators. Consult the style guide: https://github.com/nakib/elphbolt/blob/develop/STYLEGUIDE.org

File: test/test_misc.f90 [hunk 13]
```diff
@@ -334,8 +337,82 @@ program test_misc
...
+        real(r64) :: fx2
+        fx2 = sin(x)/(1.0_r64 + x**2)
+     end function fx2_el
+
+     ! Hfx2 is actual hilbert transform of fx2, is an even function
+     elemental function hfx2_el(x) result(hfx2)
+        real(r64), intent(in) :: x
+        real(r64) :: hfx2
+        hfx2 = (1/exp(1.0_r64) - cos(x))/(1.0_r64 + x**2)
+     end function hfx2_el
```
Comment (nakib) [inline test/test_misc.f90]: Why do these functions have `_el` at the end? Please remove unnecessary postfixes.

File: src/misc.f90 [hunk 14]
```diff
@@ -1653,4 +1653,45 @@ subroutine subtitle(text)
     string2print(75 - length + 1 : 75) = text
     if(this_image() == 1) write(*,'(A75)') string2print
   end subroutine subtitle
+
+  subroutine Hilbert_transform(fx, Hfx)
+     !! Does Hilbert tranform for a given function
+     !! Ref - EQ (4)3, R. Balito et. al.
+     !! "An algorithm for fast Hilbert transform of real
+     !! functions"
+     !!
```
Comment (nakib) [inline src/misc.f90]: Line 1662 still contains this random `!!`. But you marked my comment as resolved. I'm not sure why.

**Review by nakib — COMMENTED** [review]
Hi Dwaipayan,

Left my second round of review.

Best,
Nakib

File: src/misc.f90 [hunk 15]
```diff
@@ -1653,4 +1653,44 @@ subroutine subtitle(text)
     string2print(75 - length + 1 : 75) = text
     if(this_image() == 1) write(*,'(A75)') string2print
   end subroutine subtitle
+
+  subroutine Hilbert_transform(fx, Hfx)
+     !! Does Hilbert tranform for a given function
+     !! Ref - EQ (4)3, R. Balito et. al.
+     !! "An algorithm for fast Hilbert transform of real
+     !! functions"
```
Comment (nakib) [inline src/misc.f90]: I don't understand why you need a new line just for one word.

File: test/test_misc.f90 [hunk 16]
```diff
@@ -334,8 +337,78 @@ program test_misc
...
+     ! fx2 is an odd function
+     pure elemental real(r64) function fx2(x)
+        real(r64), intent(in) :: x
+        fx2 = sin(x)/(1.0_r64 + x**2)
+     end function fx2
+
+     ! Hfx2 is actual hilbert transform of fx2, is an even function
+     pure elemental real(r64) function hfx2(x)
+        real(r64), intent(in) :: x
+        hfx2 = (1/exp(1.0_r64) - cos(x))/(1.0_r64 + x**2)
```
Comment (nakib) [inline test/test_misc.f90]: Did you actually check if `1/exp(1.0_r64)` has the same precision as `1.0_r64/exp(1.0_r64)`? Why not simply write `exp(-1.0_r64)`?

**Review by nakib — CHANGES_REQUESTED** [review]

**Review by nakib — APPROVED** [review]

**Pull request merged**